### PR TITLE
Improve CPU Col2im speed and reduce output storage

### PR DESCRIPTION
**Pull request by Hosi121**
> This PR was prepared with help from an AI assistant.
>
> ## Approved by
>
> @frost-intel welcomed this PR in [#197789](https://github.com/pytorch/pytorch/issues/197789#issuecomment-5765799655).
>
> ## Summary
>
> CPU Col2im checks width bounds for each input value and processes channels in one thread. This change computes the valid width range before the row loop and uses `at::parallel_for` for separate output planes. For contiguous Fold outputs, the parallel range includes the batch. Tiles that cover the output without overlap skip the output clear and write each value once. The addition order and signed-zero behavior stay the same.
>
> Fold also allocates output storage at the input size before it resizes the output. The storage can retain much more space than the output needs. This change starts with an empty tensor. For FP32, 16 output channels, a 64x64 image, and a 3x3 kernel with stride 1 and padding 1, output storage falls from 2,359,296 to 262,144 bytes. This is tensor storage, not process RSS. No extra work tensor is added.
>
> ## Measurements
>
> Both PyTorch versions were built with the same CPU configuration. The baseline is main at `6b5c7ad181277ba21aa48b1dc03786ab95f68ad4`. The host is an Intel Core Ultra 7 255H with WSL2, GCC 13.3, Release mode, and OpenMP. Each value is the median of three process medians. The order was baseline/candidate, candidate/baseline, then baseline/candidate. CPU affinity was set for one or four threads.
>
> FP32 times in microseconds. Shapes give `N x C x H x W` for the image output.
>
> | Operation and image shape | 1 thread: main -> PR | 4 threads: main -> PR |
> |---|---:|---:|
> | Fold, 1x16x368x368, kernel/stride 2x2 | 1622.43 -> 399.87 | 1662.73 -> 102.41 |
> | Fold, 1x16x256x384, kernel/stride 2x3 | 1209.52 -> 501.53 | 1247.47 -> 145.86 |
> | Fold, 1x16x64x64, kernel 3x3, stride 1, padding 1 | 356.68 -> 63.79 | 353.45 -> 21.57 |
> | Fold, 8x1x128x128, kernel 3x3, stride 1, padding 1 | 664.67 -> 145.14 | 669.69 -> 40.99 |
> | Fold, 32x1x16x16, kernel 3x3, stride 1, padding 1 | 54.40 -> 16.17 | 54.35 -> 8.60 |
> | Unfold backward, 8x1x128x128, kernel 3x3, padding 1 | 689.23 -> 166.65 | 691.15 -> 51.91 |
>
> An empty batch took 1.73 -> 1.74 us with one thread. Native ConvTranspose2d in channels-last format took 119.32 -> 119.51 us for batch 4 with four threads; the process times overlapped. These small differences do not establish a speed change. The native ConvTranspose2d check used MKLDNN disabled. No full-model, ARM, or GPU speed result is claimed.
>
> To measure the rectangular Fold case on either build:
>
> ```python
> import torch
> from torch.utils.benchmark import Timer
>
> x = torch.randn(1, 16 * 2 * 3, 128 * 128)
> for threads in (1, 4):
>     timer = Timer(
>         "torch.nn.functional.fold(x, (256, 384), (2, 3), stride=(2, 3))",
>         globals={"torch": torch, "x": x},
>         num_threads=threads,
>     )
>     print(threads, timer.blocked_autorange(min_run_time=1))
> ```
>
> ## Tests
>
> The two builds produced equal output and gradient bits in 372 cases. The checks cover seven data types, one and four threads, noncontiguous tensors, output guards, special values, Unfold backward, and native ConvTranspose2d forward and backward.
>
> This PR adds 420 test cases to `TestNNDeviceType`, including an independent Fold reference, Unfold gradients, output batch strides, signed zero, and output storage. The commands below passed: 430 tests passed and six were skipped. The full PyTorch test suite was not run.
>
> ```bash
> PYTORCH_TESTING_DEVICE_ONLY_FOR=cpu python test/test_nn.py -k fold
> PYTORCH_TESTING_DEVICE_ONLY_FOR=cpu python test/nn/test_convolution.py -k ConvTranspose2d
> PYTORCH_TESTING_DEVICE_ONLY_FOR=cpu python test/nn/test_convolution.py -k slow_conv_dilated2d
> ```
>
> Separate ASan/UBSan probes covered the CPU entry point and kernel in 956 cases. PyTorch itself was not built with sanitizers.
>
> ## Checklist
>
> - [x] Lint passed (`spin quicklint` and `lintrunner -a --revision origin/main`).
> - [x] Tests added.
> - [x] Benchmark results included.
> - [x] No user documentation change is needed.
>
> ## BC-breaking?
>
> No. The API and numerical behavior stay the same. The output tensor retains less unused storage.

### CODE REVIEW

File: aten/src/ATen/native/im2col.h [hunk 1]
```diff
@@ -127,23 +174,77 @@ static void col2im(
...
+  };
+  constexpr int64_t grain_size = 32768;
+  // Limit the estimate when most columns contain padding.
+  const int64_t work_size = std::max(
+      image_size, std::min(image_size, output_height * output_width) * kernel_h * kernel_w);
+  if (!is_channels_last && channels > 1 && work_size * channels > grain_size) {
+    at::parallel_for(0, channels, std::max<int64_t>(1, grain_size / work_size), run);
+  } else {
+    run(0, channels);
+  }
```
Comment (frost-intel) [inline aten/src/ATen/native/im2col.h]: Though this is correct at the moment, it seems quite fragile. Someone who came along and edited the heuristic could introduce a correctness bug. How about something like
```C
if (is_channels_last) {
  col2im_slice(data_col, channels, /*...*/, data_im, /*is_channels_last=*/true);
  return;
}
at::parallel_for(0, channels, std::max<int64_t>(1, grain_size / work_size), run);
```
This would allow you to remove the `is_channel_last` parameter from `run/col2im_slice` entirely.

File: aten/src/ATen/native/im2col.h [hunk 2]
```diff
@@ -127,23 +174,77 @@ static void col2im(
...
+  const int64_t image_size = height * width;
+  const int64_t col_size = kernel_h * kernel_w * output_height * output_width;
+  const auto run = [&](int64_t begin, int64_t end) {
+    col2im_slice(
+        data_col + begin * col_size, end - begin, height, width,
+        output_height, output_width, kernel_h, kernel_w, pad_h, pad_w,
+        stride_h, stride_w, dilation_h, dilation_w,
+        data_im + begin * image_size, is_channels_last);
+  };
+  constexpr int64_t grain_size = 32768;
```
Comment (frost-intel) [inline aten/src/ATen/native/im2col.h]: This value is already hardcoded in https://github.com/pytorch/pytorch/blob/3ba8558217d2b13ed5e7053693248193309a8a0b/aten/src/ATen/TensorIterator.h#L76 . Use that in order to prevent drift.

File: aten/src/ATen/native/im2col.h [hunk 3]
```diff
@@ -127,23 +174,77 @@ static void col2im(
...
+  // Limit the estimate when most columns contain padding.
+  const int64_t work_size = std::max(
+      image_size, std::min(image_size, output_height * output_width) * kernel_h * kernel_w);
```
Comment (frost-intel) [inline aten/src/ATen/native/im2col.h]: If `output_height * output_width > image_size` that gives `image_size * hw * kw` which is larger, not limited. 
 ```suggestion
const int64_t work_size = image_size + kernel_h * kernel_w * output_height * output_width;
```

File: aten/src/ATen/native/im2col.h [hunk 4]
```diff
@@ -127,23 +174,77 @@ static void col2im(
       }
     }
   } else {
-    for (const auto c_col : c10::irange(channels_col)) {
+    for (const auto c_col : c10::irange(0, channels_col)) {
```
Comment (frost-intel) [inline aten/src/ATen/native/im2col.h]: Revert these changes, it is a diff with no purpose.

File: aten/src/ATen/native/im2col.h [hunk 5]
```diff
@@ -101,18 +101,65 @@ static void col2im(
...
+    const int64_t col_size = output_height * output_width;
+    for (const auto c : c10::irange(0, channels)) {
+      for (const auto h : c10::irange(0, output_height)) {
+        const T* src = data_col + c * 4 * col_size + h * output_width;
+        T* dst = data_im + c * height * width + h * 2 * width;
+        for (const auto w : c10::irange(0, output_width)) {
+          // Keep the addition to zero for signed-zero behavior.
+          dst[2 * w] = T(0) + src[w];
+          dst[2 * w + 1] = T(0) + src[col_size + w];
+          dst[width + 2 * w] = T(0) + src[2 * col_size + w];
```
Comment (frost-intel) [inline aten/src/ATen/native/im2col.h]: Please provide some data to justify this specialization on a 2x2 case. How much perf does this provide compared with the default case?

File: test/test_nn.py [hunk 6]
```diff
@@ -13050,6 +13050,98 @@ def func(x):
...
+        subtest((8, 1, (64, 66), (3, 3), (1, 1), (1, 1), (1, 1)), name="batch_only"),
+        subtest((2, 1, (128, 128), (1, 1), (1, 1), (0, 0), (1, 1)), name="parallel_limit"),
+        subtest((2, 1, (128, 129), (1, 1), (1, 1), (0, 0), (1, 1)), name="above_limit"),
+        subtest((0, 2, (8, 10), (2, 2), (2, 2), (0, 0), (1, 1)), name="empty_batch"),
+        subtest((2, 1, (2, 3), (2, 3), (2, 3), (5, 7), (3, 2)), name="large_padding"),
+        subtest((1, 4, (4, 5), (3, 3), (1, 1), (1, 512), (1, 1)), name="wide_padding"),
+    ])
+    def test_fold_reference(self, device, dtype, threads, noncontiguous, case):
+        self.addCleanup(torch.set_num_threads, torch.get_num_threads())
+        torch.set_num_threads(threads)
```
Comment (frost-intel) [inline test/test_nn.py]: This could perturb unrelated tests that are run in parallel. I'd drop this and consider either a single dedicated test that uses some openMP backend to gate the test.

File: test/test_nn.py [hunk 7]
```diff
@@ -13050,6 +13050,98 @@ def func(x):
...
+        subtest((2, 3, (65, 67), (2, 2), (2, 2), (0, 0), (1, 1)), name="partial_tiles"),
+        subtest((2, 3, (64, 66), (2, 2), (3, 3), (0, 0), (1, 1)), name="gaps"),
+        subtest((2, 3, (63, 65), (2, 3), (1, 2), (2, 1), (2, 3)), name="dilation"),
+        subtest((8, 1, (64, 66), (3, 3), (1, 1), (1, 1), (1, 1)), name="batch_only"),
+        subtest((2, 1, (128, 128), (1, 1), (1, 1), (0, 0), (1, 1)), name="parallel_limit"),
+        subtest((2, 1, (128, 129), (1, 1), (1, 1), (0, 0), (1, 1)), name="above_limit"),
+        subtest((0, 2, (8, 10), (2, 2), (2, 2), (0, 0), (1, 1)), name="empty_batch"),
+        subtest((2, 1, (2, 3), (2, 3), (2, 3), (5, 7), (3, 2)), name="large_padding"),
+        subtest((1, 4, (4, 5), (3, 3), (1, 1), (1, 512), (1, 1)), name="wide_padding"),
+    ])
```
Comment (frost-intel) [inline test/test_nn.py]: Far too many tests. Consider running the full cross section of tests on just float32 and maybe 1 other size for other dtypes.

File: test/test_nn.py [hunk 8]
```diff
@@ -13050,6 +13050,98 @@ def func(x):
...
+            self.assertEqual(image.grad, expected, atol=0, rtol=0)
+
+    @onlyCPU
+    @dtypes(torch.float32, torch.float64, torch.float16, torch.bfloat16,
+            torch.complex64, torch.complex128, torch.bool)
+    def test_fold_out_batch_stride(self, device, dtype):
+        columns = torch.randint(-2, 3, (4, 9, 64 * 65), device=device).to(dtype)
+        expected = F.fold(columns, (64, 65), (3, 3), padding=1)
+        storage = torch.full((8, 1, 64, 65), 7, device=device, dtype=dtype)
+        output = storage[::2]
```
Comment (frost-intel) [inline test/test_nn.py]: This is no offset. Use `storage[1::2]`

**Review by frost-intel — CHANGES_REQUESTED** [review]
Good start. I'd like to see data on how necessary the specializations are or if you could just use the one default case. Also, a more focused set of tests would be appropriate, this number of tests is too much.

**Comment by Hosi121** [comment]
> Thank you for taking the time to review this carefully. I moved the channels-last path before the parallel code and used the shared grain size with your work estimate.
>
> The added tests are down from 420 to 44. They no longer change the thread count, and the output view now uses `storage[1::2]`.
>
> For the 2x2 case, these are the FP32 results for output shape `1x16x368x368`, with kernel and stride 2. All three versions use the same CPU entry point and parallel code. Times are in microseconds on the same host as the PR measurements.
>
> | Path | 1 thread | 4 threads |
> |---|---:|---:|
> | General loop | 1089 | 219 |
> | Generic tile loop | 633 | 146 |
> | 2x2 path | 392 | 99 |
>
> I also kept the 1x1 loop: for `1x32x64x64` with one thread, it took 10.6 us versus 15.6 us for the generic tile loop. FP16 uses the generic tile loop, which was faster in these measurements.
>
> The new NCHW loops use counters; the one-argument `irange` form was slower with GCC 13.
>
> The CPU Fold and convolution tests pass.
>
> Prepared with AI assistance.

File: aten/src/ATen/native/im2col.h (hunk 1 above)
Comment (Hosi121) [reply aten/src/ATen/native/im2col.h]: > Moved the channels-last path before the parallel code and removed the layout argument from `col2im_slice` in 5476028.

File: aten/src/ATen/native/im2col.h (hunk 2 above)
Comment (Hosi121) [reply aten/src/ATen/native/im2col.h]: > Changed this to `at::internal::GRAIN_SIZE` in 5476028.

File: aten/src/ATen/native/im2col.h (hunk 3 above)
Comment (Hosi121) [reply aten/src/ATen/native/im2col.h]: > Used your work estimate in 5476028, with a minimum of 1 to avoid division by zero.

File: aten/src/ATen/native/im2col.h (hunk 4 above)
Comment (Hosi121) [reply aten/src/ATen/native/im2col.h]: > Removed the two-argument `irange` changes. The channels-last loops use their original form. The NCHW loops use counters because the one-argument `irange` form was slower with GCC 13.

File: aten/src/ATen/native/im2col.h (hunk 5 above)
Comment (Hosi121) [reply aten/src/ATen/native/im2col.h]: > For FP32 output `1x16x368x368`, with kernel and stride 2:
>
> | Path | 1 thread | 4 threads |
> |---|---:|---:|
> | General loop | 1089 | 219 |
> | Generic tile loop | 633 | 146 |
> | 2x2 path | 392 | 99 |
>
> All three versions use the same CPU entry point and parallel code. Times are in microseconds on the same host as the PR measurements.

File: test/test_nn.py (hunk 6 above)
Comment (Hosi121) [reply test/test_nn.py]: > Removed the thread-count changes from the tests in 5476028.

File: test/test_nn.py (hunk 7 above)
Comment (Hosi121) [reply test/test_nn.py]: > Reduced the added tests from 420 to 44 in 5476028. The shape matrix now uses FP32; the other data types use one representative shape.

File: test/test_nn.py (hunk 8 above)
Comment (Hosi121) [reply test/test_nn.py]: > Changed the output view to `storage[1::2]` in 5476028.
